client/transaction-pool/src/revalidation.rs

```diff
@@ -0,0 +1,346 @@
+///
+/// Implements future and can be spawned in place or in background.
+struct RevalidationWorker<Api: ChainApi> {
+	api: Arc<Api>,
+	pool: Arc<Pool<Api>>,
+	best_block: NumberFor<Api>,
+	from_queue: mpsc::UnboundedReceiver<WorkerPayload<Api>>,
+	tick_timeout: Pin<Box<dyn Stream<Item = ()> + Send>>,
+	block_ordered: BTreeMap<NumberFor<Api>, HashSet<ExHash<Api>>>,
+	members: HashMap<ExHash<Api>, NumberFor<Api>>,
+}
+
+impl<Api: ChainApi> Unpin for RevalidationWorker<Api> {}
+
+fn interval(duration: Duration) -> impl Stream<Item=()> + Unpin {
+	unfold((), move |_| {
+		Delay::new(duration).map(|_| Some(((), ())))
+	}).map(drop)
+}
+
+/// Actual async revalidation batch.
+async fn batch_revalidate<Api: ChainApi>(
+	pool: Arc<Pool<Api>>,
+	api: Arc<Api>,
+	at: NumberFor<Api>,
+	batch: impl IntoIterator<Item=ExHash<Api>>,
+) {
+	let mut invalid_hashes = Vec::new();
+	let mut revalidated = HashMap::new();
+
+	for ext_hash in batch.into_iter() {
+		let ext = match pool.ready_transaction(&ext_hash) {
+			Some(ext) => ext,
+			None => continue,
+		};
+
+		match api.validate_transaction(&BlockId::Number(at), ext.data.clone()).await {
+			Ok(Err(TransactionValidityError::Invalid(err))) => {
+				log::debug!(target: "txpool", "[{:?}]: Removed during revalidation: {:?}", ext_hash, err);
+				invalid_hashes.push(ext_hash);
+			},
+			Ok(Err(TransactionValidityError::Unknown(err))) => {
+				// skipping unknown, they might be pushed by valid or invalid transaction
+				// when latter resubmitted.
+				log::debug!(target: "txpool", "[{:?}]: Unknown during revalidation: {:?}", ext_hash, err);
+			},
+			Ok(Ok(validity)) => {
+				let data = ext.data.clone();
+				let (_, bytes) = api.hash_and_length(&data);
+				revalidated.insert(
+					ext_hash.clone(),
+					ValidatedTransaction::Valid(
+						Transaction {
+							data: ext.data.clone(),
+							bytes,
+							hash: ext_hash.clone(),
+							priority: validity.priority,
+							requires: validity.requires,
+							provides: validity.provides,
+							propagate: validity.propagate,
+							valid_till: at
+								.saturated_into::<u64>()
+								.saturating_add(validity.longevity),
+					}),
+				);
+			},
+			Err(validation_err) => {
+				log::debug!(
+					target: "txpool",
+					"[{:?}]: Error during revalidation: {:?}. Removing.",
+					ext_hash,
+					validation_err
+				);
+				invalid_hashes.push(ext_hash);
+			}
+		}
+	}
+
+	pool.remove_invalid(&invalid_hashes);
+	pool.resubmit(revalidated);
+}
+
+impl<Api: ChainApi> RevalidationWorker<Api> {
+	fn new(
+		api: Arc<Api>,
+		pool: Arc<Pool<Api>>,
+		from_queue: mpsc::UnboundedReceiver<WorkerPayload<Api>>
+	) -> Self {
+		Self {
+			api,
+			pool,
+			from_queue,
+			tick_timeout: Box::pin(interval(BACKGROUND_REVALIDATION_INTERVAL)),
+			block_ordered: Default::default(),
+			members: Default::default(),
+			best_block: Zero::zero(),
+		}
+	}
+
+	fn prepare_batch(&mut self) -> Vec<ExHash<Api>> {
+		let mut queued_exts = Vec::new();
```

Comment:
This function seems to do what it says, but I think I'd try to implement it in a bit more straight forward way (I know it's not simple though).
My attempt below (but haven't really tested if it passes the borrow checker):

```rust
let mut queued = Vec::new();
let mut left = BACKGROUND_REVALIDATION_BATCH_SIZE;

while left > 0 {
   let first_block = self.block_ordered.iter().next().cloned() {
     Some(bn) => bn,
     None => break;
   };
   let mut block_drained = false;
   if let Some(extrinsics) = self.block_ordered.get_mut(&first_block) {
       let to_queue = extrinsics.iter().take(left).cloned().collect::<Vec<_>>();
       if to_queue.len() == extrinsics.len() {
          block_drained = true;
       } else {
           for xt in &to_queue {
              extrinsics.remove(xt);
           }
       }
       left -= to_queue.len();
       queued.extend(&mut to_queue);
   }
  
   if block_drained {
     self.block_ordered.remove(&first_block);
   }
}
```
Perhaps we should use `BTreeSet` instead of `HashSet` to allow less iteration (having `split_off` would help a bit).
The issue might be though that it's processed in an ordered fashion.

Reply:
Don't see much differences, but no problem. `first_block` probably should clone just key, not key-value.